From 28c8f3dd3e1606205d54496dacfc1cfde62052ff Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sat, 22 Jun 2024 02:26:04 +0200
Subject: [PATCH] Add height-based coinbase maturity (#5)

Dogecoin changes the coinbase maturity at block 145000, so we add an upgrade mechanism for this value.

Keep the maturity on regtest at 100 so we don't have to update all the tests, but we upgrade the maturity at 1450 on regtest to test the upgrade mechanism.

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 575d33f3582e10779d5f712f58ada739f90863e0..d85b51a29a71975b478f3e96c591775f56142a6b 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -585,6 +585,7 @@ add_library(bitcoinconsensus
 	pubkey.cpp
 	uint256.cpp
 	consensus/amount.cpp
+	consensus/consensus.cpp
 	consensus/tx_check.cpp
 )
 
@@ -849,6 +850,7 @@ if(BUILD_CHAINSTATE)
 		config.cpp
 		consensus/activation.cpp
 		consensus/amount.cpp
+		consensus/consensus.cpp
 		consensus/merkle.cpp
 		consensus/tx_check.cpp
 		consensus/tx_verify.cpp
diff --git a/src/bench/block_assemble.cpp b/src/bench/block_assemble.cpp
index 404876594a0bbfee22781a004de6a2378539306c..a7ff200b92c0e2c7e0e91003c79cd084e2b511aa 100644
--- a/src/bench/block_assemble.cpp
+++ b/src/bench/block_assemble.cpp
@@ -32,7 +32,7 @@ static void AssembleBlock(benchmark::Bench &bench) {
     // Collect some loose transactions that spend the coinbases of our mined
     // blocks
     constexpr size_t NUM_BLOCKS{200};
-    std::array<CTransactionRef, NUM_BLOCKS - COINBASE_MATURITY + 1> txs;
+    std::array<CTransactionRef, NUM_BLOCKS - REGTEST_COINBASE_MATURITY + 1> txs;
     for (size_t b = 0; b < NUM_BLOCKS; ++b) {
         CMutableTransaction tx;
         tx.vin.push_back(MineBlock(config, test_setup->m_node, P2SH_OP_TRUE));
@@ -42,7 +42,7 @@ static void AssembleBlock(benchmark::Bench &bench) {
         tx.vout.emplace_back(Amount::zero(), CScript()
                                                  << OP_RETURN
                                                  << std::vector<uint8_t>(100));
-        if (NUM_BLOCKS - b >= COINBASE_MATURITY) {
+        if (NUM_BLOCKS - b >= REGTEST_COINBASE_MATURITY) {
             txs.at(b) = MakeTransactionRef(tx);
         }
     }
diff --git a/src/bench/chained_tx.cpp b/src/bench/chained_tx.cpp
index 6738177ffb57280e12227ab6e6b5912340f5a216..f90301838cc7aa31d3c38466187a705d9d243dea 100644
--- a/src/bench/chained_tx.cpp
+++ b/src/bench/chained_tx.cpp
@@ -43,7 +43,7 @@ static std::vector<CTxIn> createUTXOs(const Config &config, size_t n,
         utxos.emplace_back(MineBlock(config, node, SCRIPT_PUB_KEY));
     }
 
-    for (size_t i = 0; i < COINBASE_MATURITY + 1; ++i) {
+    for (size_t i = 0; i < REGTEST_COINBASE_MATURITY + 1; ++i) {
         MineBlock(config, node, SCRIPT_PUB_KEY);
     }
 
diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 923aafaea814f5d570bf8089a685d4c061cb7f34..69f5b0f109191d226d24dc930be4e49c16409a6b 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -127,3 +127,7 @@ bool IsMengerEnabled(const Consensus::Params &params,
 
     return IsMengerEnabled(params, pindexPrev->GetMedianTimePast());
 }
+
+bool IsDigishieldEnabled(const Consensus::Params &params, int32_t nHeight) {
+    return nHeight >= params.digishieldHeight;
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index 2c9f724d07a62c839e1ea450fdca21a2d0169070..bb35431c4afdea63ef8798997ad7f87f70074b3e 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -52,4 +52,7 @@ bool IsCowperthwaiteEnabled(const Consensus::Params &params, int32_t nHeight);
 bool IsCowperthwaiteEnabled(const Consensus::Params &params,
                             const CBlockIndex *pindexPrev);
 
+/** Check if Dogecoin Digishield protocol upgrade has activated. */
+bool IsDigishieldEnabled(const Consensus::Params &params, int32_t nHeight);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/consensus.cpp b/src/consensus/consensus.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..bffa1fd63148e5ed6c2e49f8de01385be21123c5
--- /dev/null
+++ b/src/consensus/consensus.cpp
@@ -0,0 +1,16 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <consensus/activation.h>
+#include <consensus/consensus.h>
+
+#include <consensus/params.h>
+
+int32_t CoinbaseMaturity(const Consensus::Params &params, int32_t nHeight) {
+    if (IsDigishieldEnabled(params, nHeight)) {
+        return DIGISHIELD_COINBASE_MATURITY;
+    }
+
+    return params.initialCoinbaseMaturity;
+}
diff --git a/src/consensus/consensus.h b/src/consensus/consensus.h
index c19580a67c7e8c02c79ef774a51dfa50e3f8a85c..5e7414bb3cdbad5977f635bba612acb7113867cc 100644
--- a/src/consensus/consensus.h
+++ b/src/consensus/consensus.h
@@ -8,6 +8,10 @@
 
 #include <cstdint>
 
+namespace Consensus {
+struct Params;
+}
+
 /** 1MB */
 static const uint64_t ONE_MEGABYTE = 1000000;
 /** The maximum allowed size for a transaction, in bytes */
@@ -29,7 +33,8 @@ static const int BLOCK_MAXBYTES_MAXSIGCHECKS_RATIO = 141;
  * Coinbase transaction outputs can only be spent after this number of new
  * blocks (network rule).
  */
-static const int COINBASE_MATURITY = 100;
+static const int REGTEST_COINBASE_MATURITY = 100;
+static const int DIGISHIELD_COINBASE_MATURITY = 240;
 /** Coinbase scripts have their own script size limit. */
 static const int MAX_COINBASE_SCRIPTSIG_SIZE = 100;
 
@@ -48,4 +53,6 @@ inline uint64_t GetMaxBlockSigChecksCount(uint64_t maxBlockSize) {
     return maxBlockSize / BLOCK_MAXBYTES_MAXSIGCHECKS_RATIO;
 }
 
+int32_t CoinbaseMaturity(const Consensus::Params &params, int32_t nHeight);
+
 #endif // BITCOIN_CONSENSUS_CONSENSUS_H
diff --git a/src/consensus/params.h b/src/consensus/params.h
index 1829cbf6a4783161173e6b0903db1584f4ba09dd..288bed6c9b19c41a6e8e86728a9ca7864f81dc37 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -69,6 +69,12 @@ struct Params {
     /** Unix time used for MTP activation of 15 May 2027 12:00:00 UTC upgrade */
     int sandersActivationTime;
 
+    /** Dogecoin "Digishield" activation height */
+    int digishieldHeight;
+
+    /** Coinbase maturity before Digishield */
+    int32_t initialCoinbaseMaturity;
+
     /** Enable or disable the miner fund by default */
     bool enableMinerFund;
 
diff --git a/src/consensus/tx_verify.cpp b/src/consensus/tx_verify.cpp
index d680fba3e4ef270a08cf14f62652eb564057614d..ea9c6cb7d42f96cd5952626e83e8443a670bd249 100644
--- a/src/consensus/tx_verify.cpp
+++ b/src/consensus/tx_verify.cpp
@@ -189,7 +189,7 @@ bool SequenceLocks(const CTransaction &tx, int flags,
 namespace Consensus {
 bool CheckTxInputs(const CTransaction &tx, TxValidationState &state,
                    const CCoinsViewCache &inputs, int nSpendHeight,
-                   Amount &txfee) {
+                   Amount &txfee, const Consensus::Params &params) {
     // are the actual inputs available?
     if (!inputs.HaveInputs(tx)) {
         return state.Invalid(TxValidationResult::TX_MISSING_INPUTS,
@@ -204,8 +204,8 @@ bool CheckTxInputs(const CTransaction &tx, TxValidationState &state,
         assert(!coin.IsSpent());
 
         // If prev is coinbase, check that it's matured
-        if (coin.IsCoinBase() &&
-            nSpendHeight - coin.GetHeight() < COINBASE_MATURITY) {
+        if (coin.IsCoinBase() && nSpendHeight - (int)coin.GetHeight() <
+                                     CoinbaseMaturity(params, nSpendHeight)) {
             return state.Invalid(
                 TxValidationResult::TX_PREMATURE_SPEND,
                 "bad-txns-premature-spend-of-coinbase",
diff --git a/src/consensus/tx_verify.h b/src/consensus/tx_verify.h
index 1a8d0c076d8ce2079367dd892516a80b66676884..b467995bf6d7c598fcc496505b3f1852b97c4dac 100644
--- a/src/consensus/tx_verify.h
+++ b/src/consensus/tx_verify.h
@@ -29,7 +29,7 @@ struct Params;
  */
 bool CheckTxInputs(const CTransaction &tx, TxValidationState &state,
                    const CCoinsViewCache &inputs, int nSpendHeight,
-                   Amount &txfee);
+                   Amount &txfee, const Consensus::Params &params);
 
 } // namespace Consensus
 
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index a60894eb20619c283af895780c6c55c38d542dd7..a2920b70df0c86f8798dcdf84262ce71c5cd23e0 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -8,6 +8,7 @@
 #include <chainparamsconstants.h>
 #include <chainparamsseeds.h>
 #include <consensus/amount.h>
+#include <consensus/consensus.h>
 #include <consensus/merkle.h>
 #include <primitives/block.h>
 #include <primitives/transaction.h>
@@ -155,6 +156,10 @@ public:
         // May 15, 2027 12:00:00 UTC protocol upgrade
         consensus.sandersActivationTime = 1810382400;
 
+        // Dogecoin: Digishield activation height
+        consensus.digishieldHeight = 145000;
+        consensus.initialCoinbaseMaturity = 30;
+
         /**
          * The message start string is designed to be unlikely to occur in
          * normal data. The characters are rarely used upper ASCII, not valid as
@@ -322,6 +327,10 @@ public:
         // May 15, 2027 12:00:00 UTC protocol upgrade
         consensus.sandersActivationTime = 1810382400;
 
+        // Dogecoin: Digishield activation height
+        consensus.digishieldHeight = 145000;
+        consensus.initialCoinbaseMaturity = 30;
+
         diskMagic[0] = 0xfb;
         diskMagic[1] = 0x87;
         diskMagic[2] = 0xb5;
@@ -470,6 +479,11 @@ public:
         // May 15, 2027 12:00:00 UTC protocol upgrade
         consensus.sandersActivationTime = 1810382400;
 
+        // Digishield activation height
+        consensus.digishieldHeight = 1450;
+        // keep maturity same as Bitcoin for tests
+        consensus.initialCoinbaseMaturity = REGTEST_COINBASE_MATURITY;
+
         diskMagic[0] = 0x94;
         diskMagic[1] = 0xb1;
         diskMagic[2] = 0xca;
diff --git a/src/qt/transactiondesc.cpp b/src/qt/transactiondesc.cpp
index 61f0f5e5ef2b7a8b0ca0cfebef69bafb6e869c0b..a8cc0640347c2756b0b4b97a657daee543210e15 100644
--- a/src/qt/transactiondesc.cpp
+++ b/src/qt/transactiondesc.cpp
@@ -356,7 +356,10 @@ QString TransactionDesc::toHTML(interfaces::Node &node,
     }
 
     if (wtx.is_coinbase) {
-        quint32 numBlocksToMaturity = COINBASE_MATURITY + 1;
+        quint32 numBlocksToMaturity =
+            CoinbaseMaturity(wallet.getChainParams().GetConsensus(),
+                             wallet.wallet->GetLastBlockHeight()) +
+            1;
         strHTML +=
             "<br>" +
             tr("Generated coins must mature %1 blocks before they can be "
diff --git a/src/test/fuzz/process_message.cpp b/src/test/fuzz/process_message.cpp
index f2b9786d8c25ab1c91f05a930ea649548c384d7d..1b691fd0b273f4f8622b21f951958ce2e38ef65e 100644
--- a/src/test/fuzz/process_message.cpp
+++ b/src/test/fuzz/process_message.cpp
@@ -41,7 +41,7 @@ void initialize_process_message() {
         MakeNoLogFileContext<const TestingSetup>();
     g_setup = testing_setup.get();
 
-    for (int i = 0; i < 2 * COINBASE_MATURITY; i++) {
+    for (int i = 0; i < 2 * REGTEST_COINBASE_MATURITY; i++) {
         MineBlock(GetConfig(), g_setup->m_node, CScript() << OP_TRUE);
     }
     g_setup->m_node.validation_signals->SyncWithValidationInterfaceQueue();
diff --git a/src/test/fuzz/process_messages.cpp b/src/test/fuzz/process_messages.cpp
index 33bff69135124ff80f806982e3b94faf09acdbfd..643f88b07b18f6df0e48ffbd954a1ebc70e158c7 100644
--- a/src/test/fuzz/process_messages.cpp
+++ b/src/test/fuzz/process_messages.cpp
@@ -25,7 +25,7 @@ void initialize_process_messages() {
     static const auto testing_setup =
         MakeNoLogFileContext<const TestingSetup>();
     g_setup = testing_setup.get();
-    for (int i = 0; i < 2 * COINBASE_MATURITY; i++) {
+    for (int i = 0; i < 2 * REGTEST_COINBASE_MATURITY; i++) {
         MineBlock(GetConfig(), g_setup->m_node, CScript() << OP_TRUE);
     }
     g_setup->m_node.validation_signals->SyncWithValidationInterfaceQueue();
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 8994eb4c10741520bedae987f6056dd2ccbd16cc..5830c0de82a6a477f77b0db28fea082009d02ec6 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -309,7 +309,7 @@ TestChain100Setup::TestChain100Setup(
     coinbaseKey.Set(vchKey.begin(), vchKey.end(), true);
 
     // Generate a 100-block chain:
-    this->mineBlocks(COINBASE_MATURITY);
+    this->mineBlocks(REGTEST_COINBASE_MATURITY);
 
     {
         LOCK(::cs_main);
diff --git a/src/test/validation_block_tests.cpp b/src/test/validation_block_tests.cpp
index 30305671aa741908e5e4c112551c2ed71a2da305..4d3afe83b6d9898818b297403b401e1833a5dbfb 100644
--- a/src/test/validation_block_tests.cpp
+++ b/src/test/validation_block_tests.cpp
@@ -381,7 +381,7 @@ BOOST_AUTO_TEST_CASE(mempool_locks_reorg) {
         }
 
         // Mature the inputs of the txs
-        for (int j = COINBASE_MATURITY; j > 0; --j) {
+        for (int j = REGTEST_COINBASE_MATURITY; j > 0; --j) {
             last_mined = GoodBlock(config, last_mined->GetHash());
             BOOST_REQUIRE(ProcessBlock(last_mined));
         }
@@ -392,7 +392,8 @@ BOOST_AUTO_TEST_CASE(mempool_locks_reorg) {
         std::vector<std::shared_ptr<const CBlock>> reorg;
         last_mined = GoodBlock(config, split_hash);
         reorg.push_back(last_mined);
-        for (size_t j = COINBASE_MATURITY + txs.size() + 1; j > 0; --j) {
+        for (size_t j = REGTEST_COINBASE_MATURITY + txs.size() + 1; j > 0;
+             --j) {
             last_mined = GoodBlock(config, last_mined->GetHash());
             reorg.push_back(last_mined);
         }
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 11d210d950e9a0f2627544ef245490003fa13106..f0ff5ece564271e7e1076553534499c119ddc5c3 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -480,8 +480,9 @@ void CTxMemPool::check(const CCoinsViewCache &active_coins_tip,
         TxValidationState dummy_state;
         Amount txfee{Amount::zero()};
         assert(!tx.IsCoinBase());
-        assert(Consensus::CheckTxInputs(tx, dummy_state, mempoolDuplicate,
-                                        spendheight, txfee));
+        assert(Consensus::CheckTxInputs(
+            tx, dummy_state, mempoolDuplicate, spendheight, txfee,
+            GetConfig().GetChainParams().GetConsensus()));
         for (const auto &input : tx.vin) {
             mempoolDuplicate.SpendCoin(input.prevout);
         }
diff --git a/src/validation.cpp b/src/validation.cpp
index 5f3bfc8d9f5cd9e3cb1483627101364de770e539..ea034f9c1dba7fa87151f62bdc9f5eb4a73df5b2 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -706,9 +706,9 @@ bool MemPoolAccept::PreChecks(ATMPArgs &args, Workspace &ws) {
 
     // The mempool holds txs for the next block, so pass height+1 to
     // CheckTxInputs
-    if (!Consensus::CheckTxInputs(tx, state, m_view,
-                                  m_active_chainstate.m_chain.Height() + 1,
-                                  ws.m_base_fees)) {
+    if (!Consensus::CheckTxInputs(
+            tx, state, m_view, m_active_chainstate.m_chain.Height() + 1,
+            ws.m_base_fees, args.m_config.GetChainParams().GetConsensus())) {
         // state filled in by CheckTxInputs
         return false;
     }
@@ -2363,7 +2363,7 @@ bool Chainstate::ConnectBlock(const CBlock &block, BlockValidationState &state,
             TxValidationState tx_state;
             if (!isCoinBase &&
                 !Consensus::CheckTxInputs(tx, tx_state, view, pindex->nHeight,
-                                          txfee)) {
+                                          txfee, consensusParams)) {
                 // Any transaction validation failure in ConnectBlock is a block
                 // consensus failure.
                 state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index e32c4acc267ca5d1a843bb53cbc9e2ad73d8bb73..b278e33d815bbaadba2aaf555dd07aa1bf69be9e 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -3263,10 +3263,12 @@ int CWallet::GetTxBlocksToMaturity(const CWalletTx &wtx) const {
     if (!wtx.IsCoinBase()) {
         return 0;
     }
+    const int32_t coinbaseMaturity =
+        CoinbaseMaturity(GetChainParams().GetConsensus(), GetLastBlockHeight());
     int chain_depth = GetTxDepthInMainChain(wtx);
     // coinbase tx should not be conflicted
     assert(chain_depth >= 0);
-    return std::max(0, (COINBASE_MATURITY + 1) - chain_depth);
+    return std::max(0, (coinbaseMaturity + 1) - chain_depth);
 }
 
 bool CWallet::IsTxImmatureCoinBase(const CWalletTx &wtx) const {
diff --git a/test/functional/feature_upgrade_coinbase_maturity.py b/test/functional/feature_upgrade_coinbase_maturity.py
new file mode 100644
index 0000000000000000000000000000000000000000..dbd495f1febdbfdde36eef4cf9527a3a24187e0c
--- /dev/null
+++ b/test/functional/feature_upgrade_coinbase_maturity.py
@@ -0,0 +1,197 @@
+# Copyright (c) 2024 The Bitcoin developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test coinbase maturity upgrades."""
+
+from test_framework.address import (
+    ADDRESS_ECREG_P2SH_OP_TRUE,
+    ADDRESS_ECREG_UNSPENDABLE,
+    SCRIPT_UNSPENDABLE,
+    SCRIPTSIG_OP_TRUE,
+)
+from test_framework.blocktools import create_block, create_coinbase
+from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
+from test_framework.p2p import P2PDataStore
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.txtools import pad_tx
+from test_framework.util import assert_equal, assert_raises_rpc_error
+
+
+class UpgradeCoinbaseMaturityTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [["-txindex"]]
+
+    def reconnect_p2p(self):
+        self.nodes[0].disconnect_p2ps()
+        return self.nodes[0].add_p2p_connection(P2PDataStore())
+
+    def run_test(self):
+        node = self.nodes[0]
+        peer = node.add_p2p_connection(P2PDataStore())
+
+        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        cointxid = node.getblock(coinblockhash)["tx"][0]
+
+        # Only generate 98 blocks, insufficient for maturity
+        self.generatetoaddress(node, 98, ADDRESS_ECREG_UNSPENDABLE)
+        assert_equal(node.getblockcount(), 99)
+
+        coinvalue = 5000000000
+        tx = CTransaction()
+        tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 0), SCRIPTSIG_OP_TRUE)]
+        tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(tx)
+
+        # Insufficient maturity
+        assert_raises_rpc_error(
+            -26,
+            "bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 99",
+            node.sendrawtransaction,
+            tx.serialize().hex(),
+        )
+
+        # Generate 1 more block, coins are now mature
+        self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)
+        assert_equal(node.getblockcount(), 100)
+        node.sendrawtransaction(tx.serialize().hex())
+
+        # Make sure tx got mined
+        minedhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)[0]
+        assert_equal(node.getrawtransaction(tx.hash, 2)["blockhash"], minedhash)
+
+        # Generate blocks until we're 241 blocks before upgrade height
+        UPGRADE_HEIGHT = 1450
+        OLD_MATURITY = 100
+        DIGISHIELD_MATURITY = 240
+        self.generatetoaddress(
+            node,
+            UPGRADE_HEIGHT - DIGISHIELD_MATURITY - OLD_MATURITY - 2,
+            ADDRESS_ECREG_UNSPENDABLE,
+        )
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - DIGISHIELD_MATURITY - 1)
+
+        # Get us two coins with 239 and 240 maturity at upgrade
+        mature240hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        mature240txid = node.getblock(mature240hash)["tx"][0]
+        mature239hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        mature239txid = node.getblock(mature239hash)["tx"][0]
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - DIGISHIELD_MATURITY + 1)
+
+        self.generatetoaddress(
+            node, DIGISHIELD_MATURITY - OLD_MATURITY - 3, ADDRESS_ECREG_UNSPENDABLE
+        )
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - OLD_MATURITY - 2)
+
+        # Generate two coins with 99 and 100 maturity one block before upgrade
+        mature100hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        mature100txid = node.getblock(mature100hash)["tx"][0]
+        mature99hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        mature99txid = node.getblock(mature99hash)["tx"][0]
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - OLD_MATURITY)
+
+        self.generatetoaddress(node, OLD_MATURITY - 2, ADDRESS_ECREG_UNSPENDABLE)
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - 2)
+
+        # After that many blocks, we had 8 halvings
+        coinvalue >>= 8
+
+        # For the mempool, we are still 1 block before upgrade.
+        # We can spend the 100 mature coin fine, ...
+        spend100tx = CTransaction()
+        spend100tx.vin = [
+            CTxIn(COutPoint(int(mature100txid, 16), 0), SCRIPTSIG_OP_TRUE)
+        ]
+        spend100tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(spend100tx)
+        node.sendrawtransaction(spend100tx.serialize().hex())
+
+        # ...but not the 99 mature coin
+        spend99tx = CTransaction()
+        spend99tx.vin = [CTxIn(COutPoint(int(mature99txid, 16), 0), SCRIPTSIG_OP_TRUE)]
+        spend99tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(spend99tx)
+        assert_raises_rpc_error(
+            -26,
+            "bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 99",
+            node.sendrawtransaction,
+            spend99tx.serialize().hex(),
+        )
+
+        # Can't mine the 99 mature coin spend, ...
+        block = create_block(
+            int(node.getbestblockhash(), 16), create_coinbase(node.getblockcount() + 1)
+        )
+        block.nVersion = 4
+        block.vtx += [spend99tx]
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        peer.send_blocks_and_test(
+            [block],
+            node,
+            success=False,
+            reject_reason="bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 99",
+            expect_disconnect=True,
+            timeout=5,
+        )
+        peer = self.reconnect_p2p()
+
+        # ... but we can mine the 100 mature coin spend
+        # Also, this activates the upgrade
+        minedhash1 = self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)[0]
+        assert_equal(
+            node.getrawtransaction(spend100tx.hash, 2)["blockhash"], minedhash1
+        )
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - 1)
+
+        # For the mempool the upgrade is now activated, so we can't spend a 239 mature coin
+        spend239tx = CTransaction()
+        spend239tx.vin = [
+            CTxIn(COutPoint(int(mature239txid, 16), 0), SCRIPTSIG_OP_TRUE)
+        ]
+        spend239tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(spend239tx)
+        assert_raises_rpc_error(
+            -26,
+            "bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 239",
+            node.sendrawtransaction,
+            spend239tx.serialize().hex(),
+        )
+
+        # But, we can spend the 240 mature coin
+        spend240tx = CTransaction()
+        spend240tx.vin = [
+            CTxIn(COutPoint(int(mature240txid, 16), 0), SCRIPTSIG_OP_TRUE)
+        ]
+        spend240tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(spend240tx)
+        node.sendrawtransaction(spend240tx.serialize().hex())
+
+        # Can't mine the 239 mature coin spend, ...
+        block = create_block(
+            int(node.getbestblockhash(), 16), create_coinbase(node.getblockcount() + 1)
+        )
+        block.nVersion = 4
+        block.vtx += [spend239tx]
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        peer.send_blocks_and_test(
+            [block],
+            node,
+            success=False,
+            reject_reason="bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 239",
+            expect_disconnect=True,
+        )
+        peer = self.reconnect_p2p()
+
+        # ... but we can mine the 240 mature coin spend
+        minedhash2 = self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)[0]
+        assert_equal(
+            node.getrawtransaction(spend240tx.hash, 2)["blockhash"], minedhash2
+        )
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT)
+
+
+if __name__ == "__main__":
+    UpgradeCoinbaseMaturityTest().main()
